File: packages/shared-types/src/types/agent.ts
import type { BaseEntity, UUID } from './domain';

export type AgentStatus = 'idle' | 'running' | 'paused' | 'error' | 'disabled';

export type AgentCategory = 'ingestion' | 'memory' | 'analysis' | 'retrieval' | 'action' | 'scheduling' | 'communication';

export interface Agent extends BaseEntity {
  name: string;
  description: string;
  category: AgentCategory;
  status: AgentStatus;
  version: string;
  config: AgentConfig;
  capabilities: string[];
  permissions: AgentPermissions;
  schedule?: CronSchedule;
}

export interface AgentConfig {
  model: string;
  temperature: number;
  maxTokens: number;
  tools: string[];
  memory: AgentMemoryConfig;
  promptTemplate?: string;
  rateLimit?: RateLimit;
}

export interface AgentMemoryConfig {
  enabled: boolean;
  shortTermSize: number;
  longTermEnabled: boolean;
  contextWindow: number;
}

export interface AgentPermissions {
  allowedTools: string[];
  allowedDataSources: string[];
  maxMemoryAccess: 'read' | 'write' | 'admin';
  restrictedTopics: string[];
}

export interface RateLimit {
  requestsPerMinute: number;
  requestsPerHour: number;
  concurrentLimit: number;
}

export interface CronSchedule {
  enabled: boolean;
  expression: string;
  timezone: string;
}

export interface AgentExecution {
  id: UUID;
  agentId: UUID;
  status: ExecutionStatus;
  input: Record<string, unknown>;
  output?: Record<string, unknown>;
  error?: string;
  startedAt: string;
  completedAt?: string;
  duration?: number;
  tokensUsed?: number;
  cost?: number;
}

export type ExecutionStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled' | 'timeout';

export interface AgentMessage {
  role: 'system' | 'user' | 'assistant' | 'tool';
  content: string;
  toolCalls?: ToolCall[];
  toolCallId?: string;
  metadata?: Record<string, unknown>;
}

export interface ToolCall {
  id: string;
  toolName: string;
  arguments: Record<string, unknown>;
  result?: unknown;
}

File: apps/ai-service/src/app/schemas/agent.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class AgentResponse(BaseModel):
    id: uuid.UUID
    name: str
    description: str | None = None
    category: str | None = None
    status: str
    version: str
    config: dict[str, Any]
    capabilities: list[str] | None = None
    permissions: dict[str, Any] | None = None
    tenant_id: str | None = None
    user_id: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AgentExecute(BaseModel):
    input: dict[str, Any] = Field(default_factory=dict)
    stream: bool = False


class ExecutionResponse(BaseModel):
    id: uuid.UUID
    agent_id: uuid.UUID
    status: str
    input: dict[str, Any] | None = None
    output: dict[str, Any] | None = None
    error: str | None = None
    tokens_used: int | None = None
    cost: float | None = None
    duration_ms: int | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    tenant_id: str | None = None
    user_id: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class ExecutionStatus(BaseModel):
    id: uuid.UUID
    agent_id: uuid.UUID
    status: str
    started_at: datetime | None = None
    completed_at: datetime | None = None
